### DISCUSSION TITLE
How to debug the C++ code of sherpa-onnx?

### DISCUSSION THREAD

**Discussion by OmarAlJabary** [k2-fsa/sherpa-onnx#1260] [Q&A]
Hi everyone,
I'm trying to debug the C++ sherpa-onnx code, and have difficulties printing the content of all "unique_ptr" objects, and there are plenty of them in the code (virtually every "important" object is referenced via a unique_ptr).
For example, I'm debugging the sherpa-onnx/c-api-examples/decode-file-c-api executable, and specifically focus on the function SherpaOnnxIsOnlineStreamReady.
When I step into it (after some additional step-in operations), I get to the code located here: sherpa-onnx/sherpa-onnx/csrc/online-recognizer.cc:219
bool OnlineRecognizer::IsReady(OnlineStream *s) const {
  return impl_->IsReady(s);
}
However, when I try to print the content of s or impl_ via gdb (or lldb, it doesn't matter), I get the following:
219       return impl_->IsReady(s);
(gdb) print s
$3 = (struct OnlineStream *) 0x555555ddaa70
(gdb) print *s
$4 = {impl_ = {_M_t = {_M_t = {< No data fields >}}}}
(gdb) print impl_
No symbol "impl_" in current context.

I compiled the package as follows:
git clone https://github.com/k2-fsa/sherpa-onnx
cd sherpa-onnx
mkdir build
cd build

cmake \
  -DCMAKE_BUILD_TYPE=Debug \
  -DSHERPA_ONNX_ENABLE_PYTHON=ON \
  -DBUILD_SHARED_LIBS=ON \
  -DSHERPA_ONNX_ENABLE_CHECK=OFF \
  -DSHERPA_ONNX_ENABLE_PORTAUDIO=OFF \
  -DSHERPA_ONNX_ENABLE_C_API=ON \
  -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF \
  -DSHERPA_ONNX_ENABLE_GPU=ON \
  ..

make -j

Does anyone have an idea, how can i print in gdb the content of objects referenced via unique_ptr? I tried looking over the internet, but didn't find a solution.
Maybe there is a better debugger that can do better?
Thanks!
Omar.

**Comment by csukuangfj** [comment]
Please build a debug version of sherpa-onnx.

You can pass
```
-DCMAKE_BUILD_TYPE=Debug
```
to cmake.

By the way, there is nothing special with sherpa-onnx. Everything is open-sourced. It is just plain C++.

**Reply by OmarAlJabary** [reply]
Thanks, but I already compiled it with this flag. It still doesn't show the unique_ptr.
And you're right, its really a problem of debugging C++, but in other C++ projects the behavior of showing unique_ptr objects was a bit different, and even though you need to work a bit harder to print them, it is possible, and there are guides in the internet for how to do it.
But in this project, I get completely zero information from printing them, and the guides I found in the internet don't help.

**Comment by csukuangfj** [comment]
> I compiled the package as follows:


Could you show the build logs?

Just want to check you are indeed using a debug build.

**Reply by csukuangfj** [reply]
> > I compiled the package as follows:
> 
> Could you show the build logs?
> 
> Just want to check you are indeed using a debug build.

Please answer this question.

**Comment by csukuangfj** [comment]
Also, please show the file size of the binary you are debugging.

**Comment by yuyun2000** [comment]
wow new baby

**Comment by OmarAlJabary** [comment]
Sure, the binary I'm debugging is "sherpa-onnx/c-api-examples/decode-file-c-api". Its 24KB, but it's just a test program, and it calls for other shared libraries.
I attach the log, which was obtained from "make -j 2> make.log", although I doubt you can see here if I used the debug build.
I want to make emphasize that I do see other variables, which are not unique pointers. And I do see the code itself while debugging.

**Reply by csukuangfj** [reply]
Please post the cmake + make output

**Reply by OmarAlJabary** [reply]
Sure, here they are. I ran the following commands to generate the logs:
cmake \
>   -DCMAKE_BUILD_TYPE=Debug \
  -DSHERPA_ONNX_ENABLE_PORTAUDIO=OFF \
>   -DSHERPA_ONNX_ENABLE_PYTHON=ON \
>   -DBUILD_SHARED_LIBS=ON \
>   -DSHERPA_ONNX_ENABLE_CHECK=OFF \
>   -DSHERPA_ONNX_ENABLE_PORTAUDIO=OFF \
>   -DSHERPA_ONNX_ENABLE_C_API=ON \
>   -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF \
>   -DSHERPA_ONNX_ENABLE_GPU=ON \
>   .. > cmake1.log 2> cmake2.log

make -j > make1.log 2> make2.log

**Comment by OmarAlJabary** [comment]
Did you guys had a different experience with debugging the code and printing the unique pointers? Did you use GDB or any other debugger?
Thanks!

**Comment by csukuangfj** [comment]
@OmarAlJabary 

Could you try
https://github.com/k2-fsa/sherpa-onnx/pull/1261

It should work now.

```
(lldb) p recognizer
(const SherpaOnnxOfflineRecognizer *) $0 = 0x0000600000004010
(lldb) p recognizer->impl
(const std::unique_ptr<sherpa_onnx::OfflineRecognizer, std::default_delete<sherpa_onnx::OfflineRecognizer> >) $1 = sherpa_onnx::OfflineRecognizer @ 0x0000600000004020 {
  __value_ = 0x0000600000004020
}
(lldb) p stream
(const SherpaOnnxOfflineStream *) $2 = 0x0000600000019c90
(lldb) p stream->impl
(const std::unique_ptr<sherpa_onnx::OfflineStream, std::default_delete<sherpa_onnx::OfflineStream> >) $3 = sherpa_onnx::OfflineStream @ 0x000060000001b4f0 {
  __value_ = 0x000060000001b4f0
```

**Reply by OmarAlJabary** [reply]
Thanks, it helped!
